cpp/common/src/codingstandards/cpp/types/Compatible.qll

```diff
@@ -204,79 +208,204 @@ signature module TypeEquivalenceSig {
  * To use this module, define a `TypeEquivalenceSig` module and implement a subset of `Type` that
  * selects the relevant root types to be considered. Then use the predicate `equalTypes(a, b)`.
+ * Note that `equalTypes(a, b)` only holds if `interestedIn(a, b)` holds. A type is always
+ * considered to be equal to itself, and this module does not support configurations that declare
+ * otherwise. Additionally, `interestedIn(a, b)` implies `interestedIn(b, a)`.
+ *
+ * This module will recursively select pairs of types to be compared. For instance, if
+ * `interestedInEquality(a, b)` holds, then types `a` and `b` will be compared. If
+ * `Config::equalPointerTypes(a, b, true)` holds, then the pointed-to types of `a` and `b` will be
+ * compared. However, if `Config::equalPointerTypes(a, b, false)` holds, then `a` and `b` will be
+ * compared, but their pointed-to types will not. Similarly, inner types will not be compared if
+ * `Config::overrideTypeComparison(a, b, _)` holds. For detail, see the module predicates
+ * `recurses` and `compares`.
  */
-module TypeEquivalence<TypeEquivalenceSig Config, TypeSubset T> {
+module TypeEquivalence<TypeEquivalenceSig Config, interestedInEquality/2 interestedIn> {
   /**
-   * Check whether two types are equivalent, as defined by the `TypeEquivalenceSig` module.
+   * Performance related predicate to force top down rather than bottom up evaluation of type
+   * equivalence.
+   *
+   * This interoperates with the predicate `recurses` to find types that will be compared, along
+   * with the inner types of those types that will be compared. See `recurses` for cases where this
+   * algorithm will or will not recurse. We still need to know which types are compared, even if
+   * we do not recurse on them, in order to properly constrain `equalTypes(x, y)` to hold for types
+   * such as leaf types, where we do not recurse during comparison.
+   *
+   * At each stage of recursion, we specify `pragma[only_bind_into]` to ensure that the
+   * prior `recurses` results are considered first in the pipeline.
    */
-  predicate equalTypes(RelevantType t1, RelevantType t2) {
-    if Config::overrideTypeComparison(t1, t2, _)
-    then Config::overrideTypeComparison(t1, t2, true)
-    else
-      if t1 instanceof TypedefType and Config::resolveTypedefs()
-      then equalTypes(t1.(TypedefType).getBaseType(), t2)
-      else
-        if t2 instanceof TypedefType and Config::resolveTypedefs()
-        then equalTypes(t1, t2.(TypedefType).getBaseType())
-        else (
-          not t1 instanceof DerivedType and
-          not t2 instanceof DerivedType and
-          not t1 instanceof TypedefType and
-          not t2 instanceof TypedefType and
-          LeafEquiv::getEquivalenceClass(t1) = LeafEquiv::getEquivalenceClass(t2)
-          or
-          equalDerivedTypes(t1, t2)
-          or
-          equalTypedefTypes(t1, t2)
-          or
-          equalFunctionTypes(t1, t2)
+  predicate compares(Type t1, Type t2) {
```

Comment:
`compares` feels like an unintuitive name for this predicate. Perhaps `transitiveInterestedIn` or `nestedInterestedIn`? Or `equalityCandidate`?